docs/design/remerge/rfc.md

```diff
@@ -0,0 +1,938 @@
+      `own_guid` strategy, follow the description listed in the schema format
+      document under "Merge Strategies", noting that this is the two way merge
+      case.
+
+    - For `untyped_map`: The maps are merged directly, breaking ties in favor of
+      the more recently modified.
+        - if `prefer_deletions` is true, any field represented by a tombstone in
+          either side is a tombstone in the output
+
+    - For `record_set`:
+        - The result is the set union of the two, with deletions preferred if
+          `prefer_deletion` is true
+
+5. The "merged delta" is applied to the mirror record to produce the new mirror
+   record which will be uploaded to the server.
+
+    - This record will have a vector clock that is a descendent of the local,
+      mirror, and incoming remote clocks.
+
+## New metadata records:
+
+Some additional per-collection meta-information is required to make remerge
+work.
+
+They are stored at `meta-$collection_name/blah`. This doesn't allow for
+transactional updates at the same time as the records, but in practice so long
+as the schema is always uploaded prior to uploading the records, this should be
+fine.
+
+TODO: should this be `meta/$collection_name:blah` or similar?
+
+#### `meta-$collection_name/schema`
+
+This stores the most recent schema record. See [the schema format](schema-format.md)
+reference for detailed information on its layout.
+
+#### `meta-$collection_name/client_info`
+
+Information about clients. An object with a single field currently, but possibly
+more in the future (the library must make an effort to not drop fields it does
+not understand when updating this record). The only field is `"clients"`, which
+is an array of records, each with the following properties
+
+- `"id"`: A unique ID generated on DB creation. Unrelated to any sort of current
+  client ID. Discussed in the section on counters/consistency. This is a string.
+
+    - It's illegal for this to be duplicated. If that happens, the `client_info`
+      record is considered corrupted and is discarded.
+
+- `"native_schema_version"`: This clients "native" schema version for this collection.
+    - This is a semver version string.
+    - This is the latest version it was told to use locally, even if in practice
+      it uses a more up to date schema it fetched. This is effectively the
+      version that the code adding records understands
+
+- `"local_schema_version"`: The latest version of the schema that this client
+  understands.
+    - This is also a semver version string.
+
+- `"last_sync"`: The most recent X-Weave-Timestamp (as returned by e.g. the
+  fetch to `info/collections` we do before the sync or something). This is for
+  expiring records from this list.
+
+## Migrations
+
+There are several ways you might want to do to evolve your schema, but they boil
+down to two types:
+
+Changes that have a migration path (compatible migrations), and those that do
+not (incompatible migrations).
```

Comment:
I think there's a lot of complexity here that isn't really discussed. The document seems to be talking about changing data formats, but actually I think it's talking about changing client behavior when writing data. It's easy to add a new field to a collection -- this is a compatible migration. What about adding a new field that has to be updated at the same time as some other field? This seems to me to be an incompatible migration.